### PR TITLE
Directly write snapshots to the correct location

### LINKED ISSUE #14045: No longer move/copy snapshots
## Description

Right now when we take snapshots we do that into a pending directory. If the snapshot is valid, meaning we reach a certain commit index OR we received all snapshot chunks, [we move (or copy depending on OS and filesystems) the snapshot to the valid snapshot directory](https://github.com/camunda/zeebe/blob/main/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java#L510).

Afterward, [we check the snapshot checksum](https://github.com/camunda/zeebe/blob/main/snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java#L517) (for that we have to read the whole snapshot again). This is mostly due to the case that we might fall back to copy and some error appears during copying. Important to note is that we already validated the checksum on receiving each chunk and the complete snapshot.

What if we don't copy/move the snapshot from the pending, instead we directly write to the correct location? For that, we need a different marker when snapshots are valid. This could be a separate file, which marks the snapshot either as invalid (and removing it would mark it as valid) OR the existence of that file could mark it as valid. An example could be as soon as the checksum file exists the snapshot can be seen as valid.

### Impact

This would mean we have to add some more complexity in reading and detecting valid snapshots, but in contrast, we wouldn't need to copy/move snapshots (interesting for large states) and potentially don't need to recheck the checksums again.

### Context

When working on https://github.com/camunda/zeebe/issues/13977 and https://github.com/camunda/product-hub/issues/1480 we (@npepinpe and I) discussed several issues and potential improvements regarding snapshot replication and taking snapshots, etc.. One thing that came up was that we create checksums of the whole snapshot, of each chunk, and when persisting we recreate such checksums (re-read the snapshots, etc.)

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

> Note: 
> This solves the problem of having replicated snapshots multiple times via just deleting the previous one, instead of creating unique snapshot directories. If this is not acceptable this PR should be rejected. See related comment https://github.com/camunda/zeebe/issues/14045#issuecomment-1711710503

Instead of writing to a pending directory we directly write to the snapshot directory. This allows us to remove the move
and we already make sure that a snapshot is only valid because of an existing checksum.

If we receive multiple snapshots with the same ID we delete the previous one, if it is not persisted yet.

See related slack discussion https://camunda.slack.com/archives/C037RS2JHB8/p1694182399011709

## Related issues

closes #14045 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: snapshot/src/test/java/io/camunda/zeebe/snapshots/ReceivedSnapshotTest.java
Comment (Zelldon) [inline snapshot/src/test/java/io/camunda/zeebe/snapshots/ReceivedSnapshotTest.java:95]: ℹ️ Since we no longer write to pending we are not removing/moving them

File: snapshot/src/test/java/io/camunda/zeebe/snapshots/ReceivedSnapshotTest.java
Comment (Zelldon) [inline snapshot/src/test/java/io/camunda/zeebe/snapshots/ReceivedSnapshotTest.java:176]: ℹ️ Here I was not sure how to handle this case to be honest. I'm open for alternatives.

File: snapshot/src/test/java/io/camunda/zeebe/snapshots/ReceivedSnapshotTest.java
Comment (Zelldon) [inline snapshot/src/test/java/io/camunda/zeebe/snapshots/ReceivedSnapshotTest.java:256]: i this is exactly what we do now.

**Comment by Zelldon** [comment]
Depending on whether we agree with this approach I think we can also clean up some stuff/parts like the pending directory usage etc. We might still need to clean up them, for backward compatibility. 

Since I need to work on the other topic now I would leave this up to you whether you want to go with this approach or take a different route. But I hope it helps to achieve our goals :)

**Comment by Zelldon** [comment]
We got green light 🟢 from @deepthidevaki https://camunda.slack.com/archives/C037RS2JHB8/p1694417244304599?thread_ts=1694182399.011709&cid=C037RS2JHB8 :)

**Review by npepinpe — APPROVED** [review]
:rocket:

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java
Comment (Zelldon) [reply snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java:362]: @npepinpe I might miss it but how I would check whether it exists 🤔

Comment (Zelldon) [reply snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java:362]: Ah I guess related to your previous comment via deleteFolderIfExists

**Comment by Zelldon** [comment]
I applied some comments @npepinpe I will go ahead and merge. I think we can also improve this if we see the need later.

bors r+

**Pull request merged**
